src/AdoPat/PatClient.cs

```diff
@@ -0,0 +1,139 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.Services.DelegatedAuthorization;
+
+    /// <summary>
+    /// An abstraction over common operations with the Azure DevOps PAT Lifecycle Management REST API.
+    /// </summary>
+    public class PatClient
+    {
+        private const int PageSize = 100; // Using the maximum allowable page size allows us to reduce HTTP calls.
+        private const bool AllOrgs = false; // Azure DevOps have recommended we always set this to false.
+
+        private ITokensHttpClientWrapper client;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PatClient"/> class.
+        /// </summary>
+        /// <param name="client">Any class which implements the <see cref="ITokensHttpClientWrapper"/> interface.</param>
+        public PatClient(ITokensHttpClientWrapper client)
+        {
+            this.client = client;
+        }
+
+        /// <summary>
+        /// Creates a new PAT.
+        /// </summary>
+        /// <param name="patTokenCreateRequest">
+        /// XXX: Not documented by Visual Studio client libraries. Should we depend on this or
+        /// should this method use the corresponding primitives as arguments? Not yet clear.
+        /// </param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns>A <see cref="Task"/> returning a <see cref="PatTokenResult"/>.</returns>
+        public async Task<PatToken> CreatePatAsync(
+            PatTokenCreateRequest patTokenCreateRequest,
+            CancellationToken cancellationToken = default)
+        {
+            var patTokenResult = await this.client.CreatePatAsync(
+                patTokenCreateRequest,
+                cancellationToken: cancellationToken)
+            .ConfigureAwait(false);
+
+            if (patTokenResult.PatToken == null)
+            {
+                throw new PatClientException($"Failed to create PAT during regeneration: {patTokenResult.PatTokenError}");
+            }
+
+            return patTokenResult.PatToken;
+        }
+
+        /// <summary>
+        /// Gets all active PATs.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns>The set of active PATs.</returns>
+        public async Task<ISet<PatToken>> GetActivePatsAsync(CancellationToken cancellationToken = default)
+        {
+            // Initialize a PagedPatTokens so that we can use the continuation token
+            // in the scope of the conditional of the do while loop below. Azure
+            // DevOps will return the empty string when there are no more pages.
+            var pagedPatTokens = new PagedPatTokens(continuationToken: string.Empty, patTokens: new List<PatToken>());
+            var pats = new HashSet<PatToken>();

```

Comment:
Great question. A `PatToken` has the following fields
```
string DisplayName
DateTime ValidTo
string Scope
IList<Guid> TargetAccounts
DateTime ValidFrom
Guid AuthorizationId
string Token
```
If we leave aside the `AuthorizationId`, because the `Token` is always `null` when returned from the API after the first time it would be possible to have all the other fields be the same and have duplicates. That's quite unlikely, I think, but not impossible.

However, the `AuthorizationId` does exist and it's unique to the PAT even if all the other fields are exactly the same. ❄️ 

So, no, I don't believe there's a possibility of getting duplicate PATs as long as we've got the `AuthorizationId`.

Reply:
Hmm... thinking on this a bit more, the reason I wanted an `ISet` here was to easily compare a cached PAT with what Azure DevOps thinks the state of the world is, but because of the `null` `Token` field that comparison won't actually be perfect equality. It might be better to return an `IDictionary` keyed by the `AuthorizationId`. What do you think, @mvanchaa?

Reply:
Are you suggesting we have a key value pair? Key = `AuthorizationId` and Value = '{PatToken}`

Reply:
Yes, @shalinikhare27, that's exactly what I'm suggesting. I think it will make it easier to validate whether a token that we have in the cache is in the active tokens collections. We can compare an existing `AuthorizationId` to the keys for this dictionary.

Reply:
I made this method return an `IDictionary<Guid, PatToken>` instead of an `ISet<PatToken>`, which I think will work better in the long run.